packages/ui/src/hooks/use-api.ts

```diff
@@ -0,0 +1,37 @@
+import { useState } from 'react';
+import { HTTPError } from 'ky';
+import { RequestErrorBody } from '@logto/schemas';
+
+function useApi<T extends any[], U>(api: (...arguments_: T) => Promise<U>) {
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<RequestErrorBody | null>(null);
+  const [result, setResult] = useState<U>();
+
+  const fetch = async (...arguments_: T) => {
+    setLoading(true);
+    setError(null);
+
+    try {
+      const result = await api(...arguments_);
+      setResult(result);
+      setLoading(false);
+    } catch (error_: unknown) {
+      if (error_ instanceof HTTPError) {
+        const error = (await error_.response.json()) as RequestErrorBody;
```

Comment:
```suggestion
        const kyError = await error.response.json<RequestErrorBody>();
```

Reply:
Tried unlike ky request, the response body json method takes a generic input type:
![image](https://user-images.githubusercontent.com/36393111/130236251-b440ff1b-12ef-4014-9a6d-39ae421da06b.png)
HTTPError return body json method uses the default global response body type. Which is hard coded as Promise<any>:
![image](https://user-images.githubusercontent.com/36393111/130236455-13ea8d63-9280-47db-ae8d-38735dd47305.png)

Reply:
let me take a look

Reply:
this is so fucking strange. don't know why the removed generic type definition for `.json()`. let's keep as it is for now.

Reply:
actually, you append the code snippet into `src/include.d/dom.d.ts`:

```ts
interface Body {
  json<T>(): Promise<T>;
}
```

then generic can be used

Reply:
done thx